File: GVIM2.0-system/backend/packages/harness/gvim_v2/chemistry/ketcher_commands.py
"""Canonical Ketcher command contract for native DeerFlow chemistry artifacts."""

from __future__ import annotations

from copy import deepcopy
from typing import Any

KETCHER_COMMAND_TYPES = {
    "open_editor",
    "load_molecule",
    "load_reaction",
    "load_ket",
    "add_text",
    "layout",
    "set_zoom",
    "set_settings",
    "switch_mode",
    "clear",
}

KETCHER_STRUCTURE_COMMAND_TYPES = {"load_molecule", "load_reaction", "load_ket"}

KETCHER_COMMANDS: list[dict[str, Any]] = [
    {
        "type": "open_editor",
        "description": "Render the embedded Ketcher editor.",
    },
    {
        "type": "load_molecule",
        "description": "Load a molecule from SMILES, Molfile, or MolBlock through Ketcher.setMolecule.",
        "fields": ["source", "smiles", "canonical_smiles", "molfile", "molblock", "value"],
    },
    {
        "type": "load_reaction",
        "description": "Load a reaction from an RDKit-validated RXN block or reaction SMILES through Ketcher.setMolecule.",
        "fields": ["rxnblock", "rxnfile", "reaction", "reaction_smiles", "value"],
    },
    {
        "type": "load_ket",
        "description": "Load native Ketcher KET JSON through Ketcher.setMolecule.",
        "fields": ["ket", "source", "value"],
    },
    {
        "type": "add_text",
        "description": "Add a native KET text object to the canvas.",
        "fields": ["content", "text", "value", "position", "x", "y", "z"],
    },
    {
        "type": "layout",
        "description": "Run Ketcher's native structure layout command.",
    },
    {
        "type": "set_zoom",
        "description": "Set Ketcher's editor zoom level.",
        "fields": ["zoom", "value"],
    },
    {
        "type": "set_settings",
        "description": "Apply Ketcher editor settings through Ketcher.setSettings.",
        "fields": ["settings"],
    },
    {
        "type": "switch_mode",
        "description": "Switch Ketcher between molecule and macromolecule editing modes.",
        "fields": ["mode"],
        "allowed_modes": ["molecules", "macromolecules"],
    },
    {
        "type": "clear",
        "description": "Clear the current Ketcher canvas.",
    },
]


def _string(value: Any) -> str:
    return value.strip() if isinstance(value, str) else ""


def _mode(value: Any) -> str:
    normalized = _string(value).lower()
    if normalized in {"molecule", "molecules"}:
        return "molecules"
    if normalized in {"macromolecule", "macromolecules", "sequence"}:
        return "macromolecules"
    return ""


def normalize_ketcher_command(raw: Any) -> dict[str, Any] | None:
    """Return one canonical command or None for unsupported input."""

    if not isinstance(raw, dict):
        return None

    command = deepcopy(raw)
    raw_type = _string(command.get("type"))
    if not raw_type:
        return None

    if raw_type not in KETCHER_COMMAND_TYPES:
        return None
    if raw_type == "switch_mode":
        mode = _mode(command.get("mode"))
        if not mode:
            return None
        command["mode"] = mode
    return command


def normalize_ketcher_commands(raw_commands: Any) -> list[dict[str, Any]]:
    if not isinstance(raw_commands, list):
        return []
    commands: list[dict[str, Any]] = []
    for raw in raw_commands:
        command = normalize_ketcher_command(raw)
        if command is not None:
            commands.append(command)
    return commands


def ketcher_commands_from_payload(payload: dict[str, Any]) -> list[dict[str, Any]]:
    return normalize_ketcher_commands(payload.get("ketcher_commands"))


def first_structure_command(payload: dict[str, Any]) -> dict[str, Any]:
    for command in ketcher_commands_from_payload(payload):
        if _string(command.get("type")) in KETCHER_STRUCTURE_COMMAND_TYPES:
            return command
    return {}


def with_ketcher_commands(payload: dict[str, Any]) -> dict[str, Any]:
    """Attach the canonical KetcherCommand contract."""

    output = deepcopy(payload)
    output["ketcher_command_contract"] = "1.0"
    output["ketcher_commands"] = ketcher_commands_from_payload(output)
    return output

File: GVIM2.0-system/frontend/src/core/science/ketcher-commands.ts
export type KetcherMode = "molecules" | "macromolecules";

export type KetcherCommandType =
  | "open_editor"
  | "load_molecule"
  | "load_reaction"
  | "load_ket"
  | "add_text"
  | "layout"
  | "set_zoom"
  | "set_settings"
  | "switch_mode"
  | "clear";

export type KetcherCommand = Record<string, unknown> & {
  type: KetcherCommandType;
  mode?: KetcherMode;
};

export type KetcherPayloadSource = {
  smiles: string;
  molfile: string;
  rxnblock: string;
  ket: string;
  source: string;
};

const KETCHER_COMMAND_TYPES = new Set<string>([
  "open_editor",
  "load_molecule",
  "load_reaction",
  "load_ket",
  "add_text",
  "layout",
  "set_zoom",
  "set_settings",
  "switch_mode",
  "clear",
]);

const KETCHER_STRUCTURE_COMMAND_TYPES = new Set<KetcherCommandType>([
  "load_molecule",
  "load_reaction",
  "load_ket",
]);

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function asString(value: unknown) {
  return typeof value === "string" ? value.trim() : "";
}

function firstNonEmpty(...values: string[]) {
  return values.find((value) => value.length > 0) ?? "";
}

function normalizeMode(value: unknown): KetcherMode | null {
  const mode = asString(value).toLowerCase();
  if (mode === "molecules" || mode === "molecule") {
    return "molecules";
  }
  if (
    mode === "macromolecules" ||
    mode === "macromolecule" ||
    mode === "sequence"
  ) {
    return "macromolecules";
  }
  return null;
}

export function normalizeKetcherCommand(raw: unknown): KetcherCommand | null {
  if (!isRecord(raw)) {
    return null;
  }
  const rawType = asString(raw.type);
  if (!rawType) {
    return null;
  }

  if (!KETCHER_COMMAND_TYPES.has(rawType)) {
    return null;
  }
  if (rawType === "switch_mode") {
    const mode = normalizeMode(raw.mode);
    return mode ? { ...raw, type: "switch_mode", mode } : null;
  }
  return { ...raw, type: rawType as KetcherCommandType };
}

function normalizedCommandList(value: unknown): KetcherCommand[] {
  if (!Array.isArray(value)) {
    return [];
  }
  return value
    .map((item) => normalizeKetcherCommand(item))
    .filter((item): item is KetcherCommand => item !== null);
}

export function ketcherCommandsFromPayload(
  payload: Record<string, unknown>,
): KetcherCommand[] {
  return normalizedCommandList(payload.ketcher_commands);
}

export function isKetcherStructureCommand(command: KetcherCommand) {
  return KETCHER_STRUCTURE_COMMAND_TYPES.has(command.type);
}

export function firstKetcherStructureCommand(
  payload: Record<string, unknown>,
): KetcherCommand | null {
  return (
    ketcherCommandsFromPayload(payload).find((command) =>
      isKetcherStructureCommand(command),
    ) ?? null
  );
}

export function sourceFromKetcherCommand(command: KetcherCommand | null) {
  if (!command) {
    return "";
  }
  if (command.type === "load_reaction") {
    return firstNonEmpty(
      asString(command.rxnblock),
      asString(command.rxnfile),
      asString(command.reaction),
      asString(command.reaction_smiles),
      asString(command.value),
    );
  }
  if (command.type === "load_ket") {
    return firstNonEmpty(
      asString(command.ket),
      asString(command.source),
      asString(command.value),
    );
  }
  if (command.type === "load_molecule") {
    return firstNonEmpty(
      asString(command.source),
      asString(command.smiles),
      asString(command.canonical_smiles),
      asString(command.molfile),
      asString(command.molblock),
      asString(command.value),
    );
  }
  return "";
}

export function ketcherCommandExpectsStructure(command: KetcherCommand) {
  return isKetcherStructureCommand(command) || command.type === "add_text";
}

export function ketcherCommandMode(command: KetcherCommand) {
  return command.type === "switch_mode" ? normalizeMode(command.mode) : null;
}

export function ketcherPayloadSource(
  payload: Record<string, unknown>,
): KetcherPayloadSource {
  const current = isRecord(payload.current_structure)
    ? payload.current_structure
    : {};
  const command = firstKetcherStructureCommand(payload);
  const commandSource = sourceFromKetcherCommand(command);
  const smiles = firstNonEmpty(
    asString(payload.smiles),
    asString(payload.canonical_smiles),
    asString(current.canonical_smiles),
    asString(current.smiles),
    asString(command?.smiles),
    asString(command?.canonical_smiles),
  );
  const molfile = firstNonEmpty(
    asString(payload.molfile),
    asString(payload.molblock),
    asString(current.molfile),
    asString(current.molblock),
    asString(command?.molfile),
    asString(command?.molblock),
  );
  const rxnblock = firstNonEmpty(
    asString(payload.rxnblock),
    asString(payload.rxnfile),
    asString(command?.rxnblock),
    asString(command?.rxnfile),
  );
  const ket = firstNonEmpty(
    asString(payload.ket),
    asString(current.ket),
    asString(command?.ket),
    command?.type === "load_ket" ? asString(command.value) : "",
  );
  return {
    smiles,
    molfile,
    rxnblock,
    ket,
    source: firstNonEmpty(commandSource, rxnblock, ket, smiles, molfile),
  };
}
